{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QA model  \n",
    "## References:\n",
    "- Jacob Devlin, Ming-Wei Chang, Kenton Lee, Kristina Toutanova (24 May 2019, Google AI Language); BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding; [PDF](https://arxiv.org/pdf/1810.04805.pdf)  \n",
    "- Google Research (New March 11th, 2020: Smaller BERT Models), [GitHub](https://github.com/google-research/bert)  \n",
    "- [Question Answering System In Python Using BERT NLP](https://www.pragnakalp.com/case-study/question-answering-system-in-python-using-bert-nlp/)  \n",
    "- Jonathan Besomi, [A QA model to answer them all](https://www.kaggle.com/jonathanbesomi/a-qa-model-to-answer-them-all), on the [COVID-19 Open Research Dataset challenge](https://www.kaggle.com/allen-institute-for-ai/CORD-19-research-challenge/tasks)  \n",
    "- Ranjan Satapathy (Aug 11, 2018); Question Answering in Natural Language Processing; [Medium](https://medium.com/lingvo-masino/question-and-answering-in-natural-language-processing-part-i-168f00291856)  \n",
    "- Akshay Navalakha (May 15, 2019); NLP — Question Answering System using Deep Learning; [Medium](https://medium.com/@akshaynavalakha/nlp-question-answering-system-f05825ef35c8)  \n",
    "- The Stanford Question Answering Dataset ([SQuAD](https://rajpurkar.github.io/SQuAD-explorer/))  \n",
    "- Tweet Sentiment Extraction: Extract support phrases for sentiment labels; [Kaggle](https://www.kaggle.com/c/tweet-sentiment-extraction/discussion)  \n",
    "- Yuriy Ostapov (18 Nov 2011); Question Answering in a Natural Language: Understanding System Based on Object–Oriented Semantics; [PDF](https://arxiv.org/pdf/1111.4343.pdf) \n",
    "- Yi Yang, Wen-tau Yih and Christopher Meek((2015); WIKIQA: A Challenge Dataset for Open-Domain Question Answering; [PDF](https://www.aclweb.org/anthology/D15-1237.pdf)  \n",
    "\n",
    "## Necessary Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "from pathlib import Path, PurePath\n",
    "\n",
    "import nltk\n",
    "from nltk.corpus import stopwords\n",
    "import re\n",
    "import string\n",
    "import torch\n",
    "\n",
    "from rank_bm25 import BM25Okapi #Search engine\n",
    "\n",
    "pd.options.display.max_colwidth = 100 #Display setting to show more characters in column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the processed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>abstract</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aecbc613ebdab36753235197ffb4f35734b5ca63</td>\n",
       "      <td>Clinical and immunologic studies in identical twins discordant for systemic lupus erythematosus</td>\n",
       "      <td>Brunner, Carolyn M. et al.</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>1973-08-31</td>\n",
       "      <td>Middle-aged female identical twins, one of whom had systemic lupus erythematosus (SLE), were eva...</td>\n",
       "      <td>The patient (Fo, ) was a 58 year old mentally retarded white woman, born in a rural area of sout...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>212e990b378e8d267042753d5f9d4a64ea5e9869</td>\n",
       "      <td>Infectious diarrhea: Pathogenesis and risk factors</td>\n",
       "      <td>Cantey, J.Robert</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>1985-06-28</td>\n",
       "      <td>Our understanding of the pathogenesis of infectious, especially bacterial, diarrhea has increase...</td>\n",
       "      <td>Pathogenesis and Risk Factors J. ROBERT CANTEY, M.D. Charleston, South Carolina Our understandin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bf5d344243153d58be692ceb26f52c08e2bd2d2f</td>\n",
       "      <td>New perspectives on the pathogenesis of rheumatoid arthritis</td>\n",
       "      <td>Zvaifler, Nathan J.</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>1988-10-14</td>\n",
       "      <td>In the pathogenesis of rheumatoid arthritis, locally produced antibodies complex with an incitin...</td>\n",
       "      <td>In the pathogenesis of rheumatoid arthritis, locally produced antibodies complex with an incitin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ddd2ecf42ec86ad66072962081e1ce4594431f9c</td>\n",
       "      <td>Management of acute and chronic respiratory tract infections</td>\n",
       "      <td>Ellner, Jerrold J.</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>1988-09-16</td>\n",
       "      <td>Pharyngitis, bronchitis, and pneumonia represent the most common respiratory tract infections. W...</td>\n",
       "      <td>Respiratory Tract Infections JERROLD J. ELLNER, M.D. Cleveland, CM Pharyngitis, bronchitis, and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a55cb4e724091ced46b5e55b982a14525eea1c7e</td>\n",
       "      <td>Acute bronchitis: Results of U.S. and European trials of antibiotic therapy</td>\n",
       "      <td>Dere, Willard H.</td>\n",
       "      <td>The American Journal of Medicine</td>\n",
       "      <td>1992-06-22</td>\n",
       "      <td>Acute bronchitis, an illness frequently encountered by primary-care physicians, is an inflammati...</td>\n",
       "      <td>A cute bronchitis, an illness frequently encountered by primary-care physicians [1] , is an infl...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  aecbc613ebdab36753235197ffb4f35734b5ca63   \n",
       "1  212e990b378e8d267042753d5f9d4a64ea5e9869   \n",
       "2  bf5d344243153d58be692ceb26f52c08e2bd2d2f   \n",
       "3  ddd2ecf42ec86ad66072962081e1ce4594431f9c   \n",
       "4  a55cb4e724091ced46b5e55b982a14525eea1c7e   \n",
       "\n",
       "                                                                                             title  \\\n",
       "0  Clinical and immunologic studies in identical twins discordant for systemic lupus erythematosus   \n",
       "1                                               Infectious diarrhea: Pathogenesis and risk factors   \n",
       "2                                     New perspectives on the pathogenesis of rheumatoid arthritis   \n",
       "3                                     Management of acute and chronic respiratory tract infections   \n",
       "4                      Acute bronchitis: Results of U.S. and European trials of antibiotic therapy   \n",
       "\n",
       "                      authors                           journal publish_time  \\\n",
       "0  Brunner, Carolyn M. et al.  The American Journal of Medicine   1973-08-31   \n",
       "1            Cantey, J.Robert  The American Journal of Medicine   1985-06-28   \n",
       "2         Zvaifler, Nathan J.  The American Journal of Medicine   1988-10-14   \n",
       "3          Ellner, Jerrold J.  The American Journal of Medicine   1988-09-16   \n",
       "4            Dere, Willard H.  The American Journal of Medicine   1992-06-22   \n",
       "\n",
       "                                                                                              abstract  \\\n",
       "0  Middle-aged female identical twins, one of whom had systemic lupus erythematosus (SLE), were eva...   \n",
       "1  Our understanding of the pathogenesis of infectious, especially bacterial, diarrhea has increase...   \n",
       "2  In the pathogenesis of rheumatoid arthritis, locally produced antibodies complex with an incitin...   \n",
       "3  Pharyngitis, bronchitis, and pneumonia represent the most common respiratory tract infections. W...   \n",
       "4  Acute bronchitis, an illness frequently encountered by primary-care physicians, is an inflammati...   \n",
       "\n",
       "                                                                                             body_text  \n",
       "0  The patient (Fo, ) was a 58 year old mentally retarded white woman, born in a rural area of sout...  \n",
       "1  Pathogenesis and Risk Factors J. ROBERT CANTEY, M.D. Charleston, South Carolina Our understandin...  \n",
       "2  In the pathogenesis of rheumatoid arthritis, locally produced antibodies complex with an incitin...  \n",
       "3  Respiratory Tract Infections JERROLD J. ELLNER, M.D. Cleveland, CM Pharyngitis, bronchitis, and ...  \n",
       "4  A cute bronchitis, an illness frequently encountered by primary-care physicians [1] , is an infl...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cord19_df = pd.read_csv('../data_processed/cord19_processed.csv',\n",
    "                        dtype={'paper_id':str, 'title':str, 'authors':str, 'journal':str, 'abstract':str, 'body_text':str},\n",
    "                        parse_dates=['publish_time'])\n",
    "cord19_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing and Search Engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stop words and extension words\n",
    "english_stopwords = stopwords.words('english') \n",
    "english_stopwords.extend(['_url_','_mention_','_hashtag_','figure','unmanned',\n",
    "                          'also','use','say','subject','edu','would','say','know',\n",
    "                          'good','go','get','done','try','many','nice','thank','think',\n",
    "                          'see','rather','easy','easily','lot','lack','make','want','seem',\n",
    "                          'run','need','even','right','line','even','also','may','take','come',\n",
    "                          'year','time','hour','first','last','second','high','new','low'])\n",
    "\n",
    "# Replace contractions with their longer forms\n",
    "contraction_mapping = {\"u.s.\":\"america\", \"u.s\":\"america\", \"usa\":\"america\", \"u.k.\":\"england\", \"u.k\":\"england\", \"e-mail\":\"email\",\n",
    "                       \"can't\": \"cannot\", \"'cause\": \"because\", \"could've\": \"could have\",\"he'd\": \"he would\",\"he'll\": \"he will\", \n",
    "                       \"he's\": \"he is\", \n",
    "                       \"how'd\": \"how did\", \"how'd'y\": \"how do you\", \"how'll\": \"how will\", \"how's\": \"how is\", \"I'd\": \"I would\", \n",
    "                       \"I'd've\": \"I would have\", \"I'll\": \"I will\", \"I'll've\": \"I will have\",\"I'm\": \"I am\", \"I've\": \"I have\", \n",
    "                       \"i'd\": \"i would\", \"i'd've\": \"i would have\", \"i'll\": \"i will\",  \"i'll've\": \"i will have\",\"i'm\": \"i am\",\n",
    "                       \"i've\": \"i have\", \"it'd\": \"it would\", \"it'd've\": \"it would have\", \"it'll\": \"it will\", \n",
    "                       \"it'll've\": \"it will have\",\"let's\": \"let us\", \"ma'am\": \"madam\", \"mayn't\": \"may not\", \n",
    "                       \"might've\": \"might have\",\"mightn't've\": \"might not have\", \"must've\": \"must have\",\n",
    "                       \"mustn't've\": \"must not have\", \"needn't've\": \"need not have\",\n",
    "                       \"o'clock\": \"of the clock\", \"oughtn't\": \"ought not\", \"oughtn't've\": \"ought not have\",\n",
    "                       \"sha'n't\": \"shall not\", \"shan't've\": \"shall not have\", \"she'd\": \"she would\", \"she'd've\": \"she would have\", \n",
    "                       \"she'll\": \"she will\", \"she'll've\": \"she will have\",\n",
    "                       \"shouldn't've\": \"should not have\", \"so've\": \"so have\",\"so's\": \"so as\", \n",
    "                       \"this's\": \"this is\",\"that'd\": \"that would\", \"that'd've\": \"that would have\", \"that's\": \"that is\", \n",
    "                       \"there'd\": \"there would\", \"there'd've\": \"there would have\", \"there's\": \"there is\", \"here's\": \"here is\",\n",
    "                       \"they'd\": \"they would\", \"they'd've\": \"they would have\", \"they'll\": \"they will\", \"they'll've\": \"they will have\", \n",
    "                       \"they're\": \"they are\", \"they've\": \"they have\", \"to've\": \"to have\", \"we'd\":\"we would\", \n",
    "                       \"we'd've\": \"we would have\", \"we'll\": \"we will\", \"we'll've\": \"we will have\", \"we're\":\"we are\", \"we've\":\"we have\",\n",
    "                       \"what'll\": \"what will\", \"what'll've\": \"what will have\", \"what're\": \"what are\",\n",
    "                       \"what's\": \"what is\", \"what've\": \"what have\", \"when's\": \"when is\", \"when've\": \"when have\", \"where'd\": \"where did\",\n",
    "                       \"where's\": \"where is\", \"where've\":\"where have\", \"who'll\":\"who will\", \"who'll've\":\"who will have\", \"who's\":\"who is\", \n",
    "                       \"who've\": \"who have\", \"why's\": \"why is\", \"why've\": \"why have\", \"will've\": \"will have\",\n",
    "                       \"would've\": \"would have\", \"wouldn't've\":\"would not have\", \n",
    "                       \"y'all\": \"you all\", \"y'all'd\": \"you all would\",\"y'all'd've\": \"you all would have\",\"y'all're\": \"you all are\",\n",
    "                       \"y'all've\": \"you all have\",\"you'd've\": \"you would have\"}\n",
    "\n",
    "# URL, MENTION, HASHTAG \n",
    "giant_url_regex= 'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\\(\\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'\n",
    "mention_regex  = '@[\\w\\-]+'\n",
    "hashtag_regex  = '#[\\w\\-]+'\n",
    "space_pattern  = '\\s+'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CovidSearchEngine:\n",
    "    \"\"\"\n",
    "    Simple CovidSearchEngine.\n",
    "    \"\"\"\n",
    "    \n",
    "    def text_process(self, text):\n",
    "        \n",
    "        # Deal with URL, MENTION, HASHTAG\n",
    "        text = re.sub(space_pattern, ' ', text)\n",
    "        text = re.sub(giant_url_regex, '_URL_', text)\n",
    "        text = re.sub(mention_regex, '_MENTION_', text)\n",
    "        text = re.sub(hashtag_regex, '_HASHTAG_', text)\n",
    "        \n",
    "        # Special_character\n",
    "        text = re.sub(r\"\\/\",\" \", text)\n",
    "        text = re.sub(r\"\\^\",\" ^ \", text)\n",
    "        text = re.sub(r\"\\+\",\" + \", text)\n",
    "        text = re.sub(r\"\\-\",\" - \", text)\n",
    "        text = re.sub(r\"\\=\",\" = \", text)\n",
    "        text = re.sub(r\"\\/\",\" \", text)\n",
    "        text = re.sub(r\"\\^\",\" ^ \", text)\n",
    "        text = re.sub(r\"\\+\",\" + \", text)\n",
    "        text = re.sub(r\"\\-\",\" - \", text)\n",
    "        text = re.sub(r\"\\=\",\" = \", text)\n",
    "        \n",
    "        # contraction and punctuation\n",
    "        text = text.lower()\n",
    "        text = ' '.join([contraction_mapping[t] if t in contraction_mapping else t for t in nltk.word_tokenize(text)])\n",
    "        text = text.replace(' .', '.').replace('( ', '(').replace(' )', ')')\n",
    "        text = text.translate(str.maketrans('', '', string.punctuation))\n",
    "        return text\n",
    "\n",
    "    def text_tokenize(self, text):\n",
    "        \n",
    "        # tokenize text\n",
    "        words = nltk.word_tokenize(text)\n",
    "        return list(set([word for word in words \n",
    "                         if len(word) > 2\n",
    "                         and not word in english_stopwords\n",
    "                         and not word.isnumeric() # if word.isalpha()\n",
    "                        ]))\n",
    "    \n",
    "    def preprocess(self, text):\n",
    "        # Clean and tokenize text input\n",
    "        return self.text_tokenize(self.text_process(text.lower()))\n",
    "\n",
    "    def __init__(self, corpus: pd.DataFrame):\n",
    "        self.corpus = corpus\n",
    "        self.columns = corpus.columns\n",
    "        \n",
    "        raw_search_str = self.corpus.abstract.fillna('') + ' ' \\\n",
    "                            + self.corpus.title.fillna('')\n",
    "        \n",
    "        self.index = raw_search_str.apply(self.preprocess).to_frame()\n",
    "        self.index.columns = ['terms']\n",
    "        self.index.index = self.corpus.index\n",
    "        self.bm25 = BM25Okapi(self.index.terms.tolist())\n",
    "    \n",
    "    def search(self, query, num):\n",
    "        \"\"\"\n",
    "        Return top `num` results that better match the query\n",
    "        \"\"\"\n",
    "        # obtain scores\n",
    "        search_terms = self.preprocess(query) \n",
    "        doc_scores = self.bm25.get_scores(search_terms)\n",
    "        \n",
    "        # sort by scores\n",
    "        ind = np.argsort(doc_scores)[::-1][:num] \n",
    "        \n",
    "        # select top results and returns\n",
    "        results = self.corpus.iloc[ind][self.columns]\n",
    "        results['score'] = doc_scores[ind]\n",
    "        results = results[results.score > 0]\n",
    "        return results.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 23s, sys: 514 ms, total: 1min 23s\n",
      "Wall time: 1min 24s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "cse = CovidSearchEngine(cord19_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download pre-trained QA model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'transformers'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'transformers'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "import torch\n",
    "from transformers import BertTokenizer\n",
    "from transformers import BertForQuestionAnswering\n",
    "\n",
    "# Use GPU for computation, if available \n",
    "torch_device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "\n",
    "# Use the pre-trained model to get answer to a question \n",
    "BERT_SQUAD = 'bert-large-uncased-whole-word-masking-finetuned-squad'\n",
    "\n",
    "model = BertForQuestionAnswering.from_pretrained(BERT_SQUAD)\n",
    "tokenizer = BertTokenizer.from_pretrained(BERT_SQUAD)\n",
    "\n",
    "model = model.to(torch_device)\n",
    "model.eval()\n",
    "\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Answer questions\n",
    "### Get the answers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def answer_question(question, context):\n",
    "    \n",
    "    # anser question given question and context\n",
    "    encoded_dict = tokenizer.encode_plus(\n",
    "                        question, context,\n",
    "                        add_special_tokens = True,\n",
    "                        max_length = 256,\n",
    "                        pad_to_max_length = True,\n",
    "                        return_tensors = 'pt')\n",
    "    \n",
    "    input_ids = encoded_dict['input_ids'].to(torch_device)\n",
    "    token_type_ids = encoded_dict['token_type_ids'].to(torch_device)\n",
    "    \n",
    "    start_scores, end_scores = model(input_ids, token_type_ids=token_type_ids)\n",
    "\n",
    "    all_tokens = tokenizer.convert_ids_to_tokens(input_ids[0])\n",
    "    start_index = torch.argmax(start_scores)\n",
    "    end_index = torch.argmax(end_scores)\n",
    "    \n",
    "    answer = tokenizer.convert_tokens_to_string(all_tokens[start_index:end_index+1])\n",
    "    answer = answer.replace('[CLS]', '')\n",
    "    return answer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the contexts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "NUM_CONTEXT_FOR_EACH_QUESTION = 10\n",
    "\n",
    "def get_all_context(query, num_results):\n",
    "    \"\"\"\n",
    "    Return ^num_results' papers that better match the query\n",
    "    \"\"\"\n",
    "    papers_df = cse.search(query, num_results)\n",
    "    return papers_df['abstract'].tolist()\n",
    "\n",
    "def get_all_answers(question, all_contexts):\n",
    "    \"\"\"\n",
    "    Ask the same question to all contexts (all papers)\n",
    "    \"\"\"\n",
    "    all_answers = []\n",
    "    \n",
    "    for context in all_contexts:\n",
    "        all_answers.append(answer_question(question, context))\n",
    "    return all_answers\n",
    "\n",
    "def create_output_results(question, \n",
    "                          all_contexts, \n",
    "                          all_answers, \n",
    "                          summary_answer='', \n",
    "                          summary_context=''):\n",
    "    \"\"\"\n",
    "    Return results in json format\n",
    "    \"\"\"\n",
    "    \n",
    "    def find_start_end_index_substring(context, answer):\n",
    "        search_re = re.search(re.escape(answer.lower()), context.lower())\n",
    "        if search_re:\n",
    "            return search_re.start(), search_re.end()\n",
    "        else:\n",
    "            return 0, len(context)\n",
    "        \n",
    "    output = {}\n",
    "    output['question'] = question\n",
    "    output['summary_answer'] = summary_answer\n",
    "    output['summary_context'] = summary_context\n",
    "    \n",
    "    results = []\n",
    "    for c, a in zip(all_contexts, all_answers):\n",
    "        span = {}\n",
    "        span['context'] = c\n",
    "        span['answer'] = a\n",
    "        span['start_index'], span['end_index'] = find_start_end_index_substring(c,a)\n",
    "        \n",
    "        results.append(span)\n",
    "    \n",
    "    output['results'] = results\n",
    "    \n",
    "    return output\n",
    "    \n",
    "def get_results(question, \n",
    "                summarize=False, \n",
    "                num_results=NUM_CONTEXT_FOR_EACH_QUESTION,\n",
    "                verbose=True):\n",
    "    \"\"\"\n",
    "    Get results\n",
    "    \"\"\"\n",
    "    all_contexts = get_all_context(question, num_results)\n",
    "    all_answers = get_all_answers(question, all_contexts)\n",
    "    \n",
    "    if summarize:\n",
    "        # NotImplementedYet\n",
    "        summary_answer = get_summary(all_answers)\n",
    "        summary_context = get_summary(all_contexts)\n",
    "    \n",
    "    return create_output_results(question, \n",
    "                                 all_contexts, \n",
    "                                 all_answers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tasks and Questions dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cord19_kaggle_questions = {\n",
    "    \"data\":[\n",
    "        {\n",
    "            \"task\": \"What is known about transmission, incubation, and environmental stability?\",\n",
    "            \"questions\": [\n",
    "                \"Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?\",\n",
    "                \"How long is the incubation period for the virus?\",\n",
    "                \"Can the virus be transmitted asymptomatically or during the incubation period?\",\n",
    "                \"How does weather, heat, and humidity affect the tramsmission of 2019-nCoV?\",\n",
    "                \"How long can the 2019-nCoV virus remain viable on common surfaces?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What do we know about COVID-19 risk factors?\",\n",
    "            \"questions\": [\n",
    "                \"What risk factors contribute to the severity of 2019-nCoV?\",\n",
    "                \"How does hypertension affect patients?\",\n",
    "                \"How does heart disease affect patients?\",\n",
    "                \"How does copd affect patients?\",\n",
    "                \"How does smoking affect patients?\",\n",
    "                \"How does pregnancy affect patients?\",\n",
    "                \"What is the fatality rate of 2019-nCoV?\",\n",
    "                \"What public health policies prevent or control the spread of 2019-nCoV?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What do we know about virus genetics, origin, and evolution?\",\n",
    "            \"questions\": [\n",
    "                \"Can animals transmit 2019-nCoV?\",\n",
    "                \"What animal did 2019-nCoV come from?\",\n",
    "                \"What real-time genomic tracking tools exist?\",\n",
    "                \"What geographic variations are there in the genome of 2019-nCoV?\",\n",
    "                \"What effors are being done in asia to prevent further outbreaks?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What do we know about vaccines and therapeutics?\",\n",
    "            \"questions\": [\n",
    "                \"What drugs or therapies are being investigated?\",\n",
    "                \"Are anti-inflammatory drugs recommended?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What do we know about non-pharmaceutical interventions?\",\n",
    "            \"questions\": [\n",
    "                \"Which non-pharmaceutical interventions limit tramsission?\",\n",
    "                \"What are most important barriers to compliance?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What has been published about medical care?\",\n",
    "            \"questions\": [\n",
    "                \"How does extracorporeal membrane oxygenation affect 2019-nCoV patients?\",\n",
    "                \"What telemedicine and cybercare methods are most effective?\",\n",
    "                \"How is artificial intelligence being used in real time health delivery?\",\n",
    "                \"What adjunctive or supportive methods can help patients?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"What do we know about diagnostics and surveillance?\",\n",
    "            \"questions\": [\n",
    "                \"What diagnostic tests (tools) exist or are being developed to detect 2019-nCoV?\"\n",
    "            ]\n",
    "        },\n",
    "        {\n",
    "            \"task\": \"Other interesting questions\",\n",
    "            \"questions\": [\n",
    "                \"What is the immune system response to 2019-nCoV?\",\n",
    "                \"Can personal protective equipment prevent the transmission of 2019-nCoV?\",\n",
    "                \"Can 2019-nCoV infect patients a second time?\"\n",
    "            ]\n",
    "        }\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Answering questions to task 1. ...\n",
      "Answering questions to task 2. ...\n",
      "Answering questions to task 3. ...\n",
      "Answering questions to task 4. ...\n",
      "Answering questions to task 5. ...\n",
      "Answering questions to task 6. ...\n",
      "Answering questions to task 7. ...\n",
      "Answering questions to task 8. ...\n",
      "Hey, we're okay! All the questions of the tasks are answered.\n",
      "CPU times: user 25min 14s, sys: 1min 25s, total: 26min 39s\n",
      "Wall time: 6min 43s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "all_tasks = []\n",
    "\n",
    "for i, t in enumerate(cord19_kaggle_questions['data']):\n",
    "    print(\"Answering questions to task {}. ...\".format(i+1))\n",
    "    answers_to_question = []\n",
    "    for q in t['questions']:\n",
    "            answers_to_question.append(get_results(q, verbose=False))\n",
    "    task = {}\n",
    "    task['task'] = t['task']\n",
    "    task['questions'] = answers_to_question\n",
    "    \n",
    "    all_tasks.append(task)\n",
    "\n",
    "print(\"Hey, we're okay! All the questions of the tasks are answered.\")\n",
    "cord19_answers_json = {}\n",
    "cord19_answers_json['data'] = all_tasks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To JSON format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to JSON format\n",
    "#import json\n",
    "#with open('../data_processed/cord19_answers.json', 'w') as outfile:\n",
    "#    json.dump(cord19_answers_json, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To CSV format\n",
    "#### Answer reference (authors and year of publication)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "authors_year = []\n",
    "for j in range(cord19_df.shape[0]):\n",
    "    authors_year.append(str(cord19_df.authors[j])+' ('+str(cord19_df.publish_time[j].year)+')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>task</th>\n",
       "      <th>question</th>\n",
       "      <th>context</th>\n",
       "      <th>answer</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>What is known about transmission, incubation, and environmental stability?</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?</td>\n",
       "      <td>Viral infections can be transmitted by various routes. At one extreme, airborne or droplet viral...</td>\n",
       "      <td>transmission is via aerosols , droplets , or fomites ( contaminated surfaces ) . environmental f...</td>\n",
       "      <td>Limeres Posse, Jacobo et al. (2017)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What is known about transmission, incubation, and environmental stability?</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?</td>\n",
       "      <td>Bacterial infections have a large impact on public health. Disease can occur at any body site an...</td>\n",
       "      <td>air , water , food , or living vectors</td>\n",
       "      <td>Doron, S.; Gorbach, S.L. (2008)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>What is known about transmission, incubation, and environmental stability?</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?</td>\n",
       "      <td>BACKGROUND: Respiratory infections are the leading cause of childhood deaths in Bangladesh. Prom...</td>\n",
       "      <td>close contact with a sick person ' s breath , cough droplets , or spit</td>\n",
       "      <td>Nizame, Fosiul A et al. (2011)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What is known about transmission, incubation, and environmental stability?</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?</td>\n",
       "      <td>Emerging infectious diseases (EID) and reemerging infectious diseases are increasing globally. Z...</td>\n",
       "      <td>direct contact or through food , water , and the environment</td>\n",
       "      <td>McArthur, Donna Behler (2019)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>What is known about transmission, incubation, and environmental stability?</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?</td>\n",
       "      <td>Viruses are often transmitted via food and the environment. Contamination may be controlled eith...</td>\n",
       "      <td>viruses are often transmitted via food and the environment</td>\n",
       "      <td>Cliver, Dean O. (2008)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         task  \\\n",
       "0  What is known about transmission, incubation, and environmental stability?   \n",
       "1  What is known about transmission, incubation, and environmental stability?   \n",
       "2  What is known about transmission, incubation, and environmental stability?   \n",
       "3  What is known about transmission, incubation, and environmental stability?   \n",
       "4  What is known about transmission, incubation, and environmental stability?   \n",
       "\n",
       "                                                                                      question  \\\n",
       "0  Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?   \n",
       "1  Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?   \n",
       "2  Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?   \n",
       "3  Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?   \n",
       "4  Is the virus transmitted by aerisol, droplets, food, close contact, fecal matter, or water?   \n",
       "\n",
       "                                                                                               context  \\\n",
       "0  Viral infections can be transmitted by various routes. At one extreme, airborne or droplet viral...   \n",
       "1  Bacterial infections have a large impact on public health. Disease can occur at any body site an...   \n",
       "2  BACKGROUND: Respiratory infections are the leading cause of childhood deaths in Bangladesh. Prom...   \n",
       "3  Emerging infectious diseases (EID) and reemerging infectious diseases are increasing globally. Z...   \n",
       "4  Viruses are often transmitted via food and the environment. Contamination may be controlled eith...   \n",
       "\n",
       "                                                                                                answer  \\\n",
       "0  transmission is via aerosols , droplets , or fomites ( contaminated surfaces ) . environmental f...   \n",
       "1                                                               air , water , food , or living vectors   \n",
       "2                               close contact with a sick person ' s breath , cough droplets , or spit   \n",
       "3                                         direct contact or through food , water , and the environment   \n",
       "4                                           viruses are often transmitted via food and the environment   \n",
       "\n",
       "                             reference  \n",
       "0  Limeres Posse, Jacobo et al. (2017)  \n",
       "1      Doron, S.; Gorbach, S.L. (2008)  \n",
       "2       Nizame, Fosiul A et al. (2011)  \n",
       "3        McArthur, Donna Behler (2019)  \n",
       "4               Cliver, Dean O. (2008)  "
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "j = 0\n",
    "results_df = pd.DataFrame(columns=['task','question','context','answer','reference'])\n",
    "for tasks in cord19_answers_json['data']:\n",
    "    for questions in tasks['questions']:\n",
    "        for question in questions['results']:\n",
    "            results_df.loc[j,'task']    = tasks['task']\n",
    "            results_df.loc[j,'question']= questions['question']\n",
    "            results_df.loc[j,'context'] = question['context']\n",
    "            results_df.loc[j,'answer']  = question['answer']\n",
    "            j+=1\n",
    "\n",
    "for abstract, ref in zip(cord19_df.abstract, authors_year):\n",
    "    for j, context in enumerate(results_df.context):\n",
    "        if context in abstract:\n",
    "            results_df.loc[j,'reference'] = ref\n",
    "\n",
    "results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#results_df.to_csv('../data_processed/cord19_answers.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "with open('../data_processed/cord19_answers.json', mode='r') as file:\n",
    "     cord19_answers_json = json.loads(file.read())\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "tasks = [\n",
    "    {'label': task['task'], \n",
    "     'value': task['task'],\n",
    "    } \n",
    "    for task in cord19_answers_json['data']]\n",
    "\n",
    "questions = [\n",
    "    {\n",
    "        'label': [q['question'] for q in tq['questions']],\n",
    "        'value': [q['question'] for q in tq['questions']],\n",
    "    }\n",
    "    for tq in cord19_answers_json['data']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HTML Layout Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown, Latex, HTML\n",
    "\n",
    "def layout_style():\n",
    "    style = \"\"\"\n",
    "        div {\n",
    "            color: black;\n",
    "            }\n",
    "        .single_answer {\n",
    "            border-left: 3px solid green;\n",
    "            padding-left: 10px;\n",
    "            font-family: Arial;\n",
    "            font-size: 16px;\n",
    "            color: #777777;\n",
    "            margin-left: 5px;\n",
    "            }\n",
    "        .answer{\n",
    "            color: #dc7b15;\n",
    "            }\n",
    "        .ay{\n",
    "            color: red;\n",
    "            }\n",
    "        .question_title {\n",
    "            color: darkblue;\n",
    "            display: block;\n",
    "            text-transform: none;\n",
    "            }\n",
    "        .task_title {\n",
    "            color: darkgreen;\n",
    "            }\n",
    "        div.output_scroll {\n",
    "            height: auto;\n",
    "            }\n",
    "    \"\"\"\n",
    "    return \"<style>\" + style + \"</style>\"\n",
    "\n",
    "def dm(x): display(Markdown(x))\n",
    "def dh(x): display(HTML(layout_style() + x))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Display the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_single_context(context, start_index, end_index):\n",
    "    \n",
    "    before_answer = context[:start_index]\n",
    "    answer = context[start_index:end_index]\n",
    "    after_answer = context[end_index:]\n",
    "    \n",
    "    content = before_answer + \"<span class='answer'>\" + answer + \"</span>\" + after_answer\n",
    "    \n",
    "    return dh(\"\"\"<div class=\"single_answer\">{}</div>\"\"\".format(content))\n",
    "\n",
    "def display_question_title(question):\n",
    "    return dh(\"<h2 class='question_title'>{}</h2>\".format(question)) #.capitalize()\n",
    "\n",
    "def display_all_contexts(index, question):\n",
    "    \n",
    "    def answer_not_found(context, start_index, end_index):\n",
    "        return (start_index == 0 and len(context) == end_index) or (start_index == 0 and end_index == 0)\n",
    "\n",
    "    display_question_title(str(index + 1) + \". \" + question['question'].capitalize())\n",
    "    \n",
    "    # display context\n",
    "    for i in question['results']:\n",
    "        for a, ay in zip(cord19_df.abstract, authors_year):\n",
    "            if i['context'] in a:\n",
    "                \n",
    "                if answer_not_found(i['context'], i['start_index'], i['end_index']):\n",
    "                    continue # skip not found questions\n",
    "                display_single_context(i['context']+'<br>'+'<strong>'+'<font color=black>'+ay,\n",
    "                                       i['start_index'], i['end_index'])\n",
    "\n",
    "def display_task_title(index, task):\n",
    "    task_title = \"Task \" + str(index) + \": \" + task\n",
    "    return dh(\"<h1 class='task_title'>{}</h1>\".format(task_title))\n",
    "\n",
    "def display_single_task(index, task):\n",
    "    \n",
    "    display_task_title(index, task['task'])\n",
    "    \n",
    "    for i, question in enumerate(task['questions']):\n",
    "        display_all_contexts(i, question)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a task number 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h1 class='task_title'>Task 4: What do we know about vaccines and therapeutics?</h1>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h2 class='question_title'>1. What drugs or therapies are being investigated?</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">Twenty antiviral drugs, that is about half of those that are currently approved, are formally licensed for clinical use in the treatment of human immunodeficiency virus infections (acquired immune deficiency syndrome). The others are used in the treatment of herpesvirus (e.g. herpes simplex virus, varicella zoster virus and cytomegalo virus), hepatitis B virus, hepatitis C virus or influenza virus infections. Recent endeavours have focussed on the development of <span class='answer'>improved antiviral therapies</span> for virus infections that have already proved amenable to antiviral drug treatment, as well as for virus infections for which, at present, no antiviral drugs have been formally approved (i.e. human papilloma viruses, adenoviruses, human herpesvirus type 6, poxviruses, severe acute respiratory syndrome coronavirus and hemorrhagic fever viruses).<br><strong><font color=black>De Clercq, Erik (2005)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">The dramatic increase in both the number of novel infectious agents and resistance to antimicrobial drugs has incited the need for <span class='answer'>adjunct therapies</span> in the war against infectious diseases. Exciting recent studies have demonstrated the use of antibodies in the form of intravenous immunoglobulin (IVIg) against infections. By virtue of the diverse repertoire of immunoglobulins that possess a wide spectrum of antibacterial and antiviral specificities, IVIg provides antimicrobial efficacy independently of pathogen resistance and represents a promising alternative strategy for the treatment of diseases for which a specific therapy is not yet available.<br><strong><font color=black>Bayry, Jagadeesh et al. (2004)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">ABSTRACT Viral infections play an important role in human diseases, and recent outbreaks in the advent of globalization and ease of travel have underscored their prevention as a critical issue in safeguarding public health. Despite the progress made in immunization and drug development, many viruses lack preventive vaccines and efficient antiviral therapies, which are often beset by the generation of viral escape mutants. Thus, identifying <span class='answer'>novel antiviral drugs</span> is of critical importance and natural products are an excellent source for such discoveries. In this mini-review, we summarize the antiviral effects reported for several natural products and herbal medicines.<br><strong><font color=black>Lin, Liang-Tzung et al. (2014)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h2 class='question_title'>2. Are anti-inflammatory drugs recommended?</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">Feline Chronic Gingivostomatitis Syndrome (FCGS) is a common disease in clinical practice. Among the therapeutic options available, long-acting corticosteroids are frequently used due to their anti-inflammatory and immunosuppressive properties. Although they may improve the clinical symptoms, they can lead to a progressive form of the disease that becomes refractory to treatment. Furthermore, their direct relationship with type II diabetes mellitus (DM) is well known. Consequently, these drugs are controversial and <span class='answer'>not recommended</span> for routine management of FCGS. Recombinant feline interferon-omega (rFeIFN-ω) is an immunomodulatory compound. Recently, its daily oral administration has been shown to be successful in treating refractory cases of FCGS. This case study describes two clinical cases of type II DM complicated by FCGS. Both animals were calicivirus positive and they had been previously treated with long-acting corticosteroids, which may have been the major cause of DM. The two cats were treated with glargine insulin (Lantus, starting dose 1 IU/cat twice daily (BID)), achieving remission 10 and 18 weeks later respectively. Considering the difficulty with control of FCGS in these animals, an oral daily dose of rFeIFN-ω was started as an alternative to long-acting corticosteroids. In both cats oral clinical signs gradually improved and 60 days after the start of therapy the owners reported a significant relief of pain during mastication. According to the authors’ knowledge, this is the first case report that describes the successful use of rFeIFN-ω in the management of FCGS in type II diabetic cats, in which long-acting corticosteroids are contraindicated.<br><strong><font color=black>Leal, Rodolfo O et al. (2013)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">Essential oils from various aromatic medicinal plants are highly active against some viral infections, e.g. labial herpes caused by herpes simplex virus type 1. Balm oil, tea tree oil and peppermint oil demonstrate in vitro a significant antiherpetic activity, mainly related to a direct drug-virus particle interaction, some essential oils also act directly virucidal. Interestingly, these essential oils are also highly active against acyclovir-resistant herpes simplex virus strains. In clinical studies, tea tree oil has been shown to possess antiherpetic, anti-inflammatory and pain-relieving properties, as well as to accelerate the healing process of herpes labialis. <span class='answer'>Applying diluted essential oils three to four times daily for the antiherpetic treatment of affected areas is recommended</span>. Some companies have marketed plant products, e.g. from Melissa, for the treatment of recurrent herpetic infections.<br><strong><font color=black>Schnitzler, P.; Reichling, J. (2011)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "task = int(input('Enter a task number'))\n",
    "\n",
    "display_single_task(task, cord19_answers_json['data'][task-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter a task number 8\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h1 class='task_title'>Task 8: Other interesting questions</h1>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h2 class='question_title'>1. What is the immune system response to 2019-ncov?</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">This paper focuses on the formulation of a deterministic 2019-nCov transmission model by considering the <span class='answer'>exposed and recovered populations</span> with immunity. The scenario of the simulation is depicted based on the patient zero in Malaysia. The transmission model is found to be able to predict the next confirmed case given a single case introduced in a fully susceptible population. The mathematical model is developed based on the SEIR model which has susceptible, exposed, infectious and recovered populations. The system of equations which were obtained were solved numerically and the simulation results were analyzed. The analysis includes the impact of the disease if no control is taken.<br><strong><font color=black>Jane Labadin; Boon Hao Hong (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">The 2019-nCoV infection that is caused by a novel strain of coronavirus was first detected in China in the end of December 2019 and declared a public health emergency of international concern by the World Health Organization on January 30, 2020. Community pharmacists in one of the first areas that had confirmed cases of the viral infection, Macau, joined the collaborative force in supporting the <span class='answer'>local health emergency preparedness and response arrangements</span>. This paper aimed to improve the understanding of community pharmacists’ role in case of 2019-CoV outbreak based on the practical experiences in consultation with the recommendations made by the International Pharmaceutical Federation on the Coronavirus 2019-nCoV outbreak.<br><strong><font color=black>Ung, Carolina Oi Lam (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h2 class='question_title'>2. Can personal protective equipment prevent the transmission of 2019-ncov?</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">In late December 2019 a previous unidentified coronavirus, currently named as the 2019 novel coronavirus (2019-nCoV), emerged from Wuhan, China and resulted in a formidable outbreak in many cities in China and expanding globally, including Thailand, Republic of Korea, Japan, USA, Philippines, Viet Nam, and our country (as of 2/6/2020 at least 25 countries). The disease is officially named as the Severe Specific Contagious Pneumonia (SSCP) in 1/15/2019 and is a notifiable communicable disease of the 5 category by the Taiwan CDC, the Ministry of Health. SSCP is a potential zoonotic disease with low to moderate (estimated 2-5%) mortality rate. Person-to-person transmission may occur through droplet or contact transmission and jeopardized first-line healthcare workers if <span class='answer'>lack of stringent infection control or no proper personal protective equipment available</span>. Currently, there is no definite treatment for SSCP although some drugs are under investigation. To promptly identify patients and prevent further spreading, physicians should be aware of travel or contact history for patients with compatible symptoms.<br><strong><font color=black>Wu, Yi-Chi et al. (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">ABSTRACT THE CENTERS FOR DISEASE CONTROL and Prevention has published updated guidelines for isolation precautions that outline how health care workers can prevent the transmission of infectious agents to their patients and to themselves. THE GUIDELINES RE-EMPHASIZE standard precautions, which guide clinicians in the use of appropriate personal protective equipment (PPE), <span class='answer'>depending on the expected type of exposure</span>. Respiratory hygiene/cough etiquette is incorporated into infection control practices as a new component of standard precautions. THE ARTICLE PROVIDES information on the new guidelines as well as information on newly emerging patho gens and methods to prevent disease transmission in health care settings.<br><strong><font color=black>Tarrac, Shauna Ely (2008)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">The severe acute respiratory syndrome (SARS) outbreak in Vietnam was amplified by nosocomial spread within hospital A, but no transmission was reported in hospital B, the second of two designated SARS hospitals. Our study documents lack of SARS-associated coronavirus transmission to hospital B workers, despite variable infection control measures and <span class='answer'>the use of personal protective equipment</span>.<br><strong><font color=black>Ha, Le Dang et al. (2004)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">A possible threat in the ophthalmology clinic While the 2019-nCoV transmission route is still unknown, <span class='answer'>countries have been preparing measures</span> based on past experiences with coronaviruses namely SARS-CoV and MERS-CoV. These viruses transmit primarily through droplets and other bodily secretions. In the ophthalmology practice, healthcare workers may be particularly susceptible to these infections. Firstly, ophthalmologists are extremely reliant on physical examination during patient consultation. Of particular concern is the proximity between the patient and ophthalmologist during the slit lamp microscope examination. It has been shown that droplets from a cough or sneeze can be propelled for up to 6?m [8], a range that definitely encompasses the distance between the patient and ophthalmologist. Secondly, during the SARS-CoV epidemic, clinical reports have suggested tears as a medium of infection. In a case series by Loon et al., it was shown that viral RNA of the SARS-CoV can be detected by reverse-transcription polymerase chain reaction (RT-PCR) from the tears of infected individuals [9]. While anecdotal in nature, such accounts highlight the possible infectivity of tears, a fluid which ophthalmologists and instruments come in contact on a daily basis. If true, this represents a crucial need for further development of disinfection and personal protective equipment (PPE) protocols for the ophthalmology clinic.SN - 1476-5454<br><strong><font color=black>Seah, Ivan et al. (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">In late December 2019, a previous unidentified coronavirus, currently named as the 2019 novel coronavirus#, emerged from Wuhan, China, and resulted in a formidable outbreak in many cities in China and expanded globally, including Thailand, Republic of Korea, Japan, United States, Philippines, Viet Nam, and our country (as of 2/6/2020 at least 25 countries). The disease is officially named as Coronavirus Disease-2019 (COVID-19, by WHO on February 11, 2020). It is also named as Severe Pneumonia with Novel Pathogens on January 15, 2019 by the Taiwan CDC, the Ministry of Health and is a notifiable communicable disease of the fifth category. COVID-19 is a potential zoonotic disease with low to moderate (estimated 2%-5%) mortality rate. Person-to-person transmission may occur through droplet or contact transmission and <span class='answer'>if there is a lack of stringent infection control or if no proper personal protective equipment available</span>, it may jeopardize the first-line healthcare workers. Currently, there is no definite treatment for COVID-19 although some drugs are under investigation. To promptly identify patients and prevent further spreading, physicians should be aware of the travel or contact history of the patient with compatible symptoms.<br><strong><font color=black>Wu, Y. C. et al. (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">This document outlines recommendations for physiotherapy management for COVID-19 in the acute hospital setting. It includes: recommendations for physiotherapy workforce planning and preparation; a screening tool for determining requirement for physiotherapy; and <span class='answer'>recommendations for the selection of physiotherapy treatments and personal protective equipment</span>. It is intended for use by physiotherapists and other relevant stakeholders in the acute care setting caring for adult patients with suspected and/or confirmed COVID-19.<br><strong><font color=black>Thomas, Peter et al. (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">Summary The epidemiology and virology of severe acute respiratory syndrome (SARS) have been written about many times and several guidelines on the infection control and public health measures believed necessary to control the spread of the virus have been published. However, there have been few reports of the problems that infectious disease clinicians encounter when dealing with the protean manifestations of this pathogen. This is a qualitative account of some of the issues faced by an infectious disease physician when identifying and treating patients with SARS as well as protecting other healthcare workers and patients, including: identification of the chain of contagion, early recognition of the disease in the absence of a reliable and rapid diagnostic test, <span class='answer'>appropriate use of personal protective equipment</span>, and the use of isolation to prevent super-spreading events. Many issues need to be addressed if clinicians are to be able to manage the virus should it reappear.<br><strong><font color=black>Tambyah, Paul A (2004)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><h2 class='question_title'>3. Can 2019-ncov infect patients a second time?</h2>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">Translation of a genetic codon without a cognate tRNA gene is affected by both the cognate tRNA availability and the interaction with non-cognate isoacceptor tRNAs. Moreover, two consecutive slow codons (slow di-codons) lead to a much slower translation rate. Calculating the composition of host specific slow codons and slow di-codons in the viral protein coding sequences can predict the order of viral protein synthesis rates between different virus strains. Comparison of human-specific slow codon and slow di-codon compositions in the genomes of 590 coronaviruses infect humans revealed that the protein synthetic rates of 2019 novel coronavirus (2019-nCoV) and severe acute respiratory syndrome-related coronavirus (SARS-CoV) <span class='answer'>may be much faster than other coronaviruses infect humans</span>. Analysis of host-specific slow codon and di-codon compositions provides links between viral genomic sequences and capability of virus replication in host cells that may be useful for surveillance of the transmission potential of novel viruses.<br><strong><font color=black>Yang, Chu-Wen; Chen, Mei-Fang (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        div {\n",
       "            color: black;\n",
       "            }\n",
       "        .single_answer {\n",
       "            border-left: 3px solid green;\n",
       "            padding-left: 10px;\n",
       "            font-family: Arial;\n",
       "            font-size: 16px;\n",
       "            color: #777777;\n",
       "            margin-left: 5px;\n",
       "            }\n",
       "        .answer{\n",
       "            color: #dc7b15;\n",
       "            }\n",
       "        .ay{\n",
       "            color: red;\n",
       "            }\n",
       "        .question_title {\n",
       "            color: darkblue;\n",
       "            display: block;\n",
       "            text-transform: none;\n",
       "            }\n",
       "        .task_title {\n",
       "            color: darkgreen;\n",
       "            }\n",
       "        div.output_scroll {\n",
       "            height: auto;\n",
       "            }\n",
       "    </style><div class=\"single_answer\">AbstractOver the past 20 years, several coronaviruses have crossed the species barrier into humans, causing outbreaks of severe, and often fatal, respiratory illness. Since SARS- CoV was first identified in animal markets, global viromics projects have discovered thousands of coronavirus sequences in diverse animals and geographic regions. Unfortunately, there are few tools available to functionally test these novel viruses for their ability to infect humans, which has severely hampered efforts to predict the next zoonotic viral outbreak. Here we developed an approach to rapidly screen lineage B betacoronaviruses, such as SARS-CoV and the recent 2019-nCoV, for receptor usage and their ability to infect cell types from different species. We show that host protease processing during viral entry is a significant barrier for several lineage B viruses and that bypassing this barrier allows several lineage B viruses to enter human cells through an unknown receptor. We also demonstrate how different lineage B viruses can recombine to gain entry into human cells and confirm that <span class='answer'>human ACE2 is the receptor</span> for the recently emerging 2019-nCoV.<br><strong><font color=black>Michael Letko; Vincent Munster (2020)</div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "task = int(input('Enter a task number'))\n",
    "\n",
    "display_single_task(task, cord19_answers_json['data'][task-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ra' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-66-df3e6fbac965>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'Context'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m}\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mra\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;31m#output[context]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m#html.P(output['Context'+str(j)] for j, out in enumerate(ra))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ra' is not defined"
     ]
    }
   ],
   "source": [
    "    output = {'Context'+str(j): {out} for j, out in enumerate(ra)}\n",
    "    #output[context]\n",
    "    #html.P(output['Context'+str(j)] for j, out in enumerate(ra))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "======================================="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "# =========================================================================================================\n",
    "from IPython.display import display, Markdown, Latex, HTML\n",
    "\n",
    "def layout_style():\n",
    "    style = \"\"\"\n",
    "        div {\n",
    "            color: black;\n",
    "            }\n",
    "        .single_answer {\n",
    "            border-left: 3px solid green;\n",
    "            padding-left: 10px;\n",
    "            font-family: Arial;\n",
    "            font-size: 16px;\n",
    "            color: #777777;\n",
    "            margin-left: 5px;\n",
    "            }\n",
    "        .answer{\n",
    "            color: #dc7b15;\n",
    "            }\n",
    "        .ay{\n",
    "            color: red;\n",
    "            }\n",
    "        .question_title {\n",
    "            color: darkblue;\n",
    "            display: block;\n",
    "            text-transform: none;\n",
    "            }\n",
    "        .task_title {\n",
    "            color: darkgreen;\n",
    "            }\n",
    "        div.output_scroll {\n",
    "            height: auto;\n",
    "            }\n",
    "    \"\"\"\n",
    "    return \"<style>\" + style + \"</style>\"\n",
    "\n",
    "def dm(x): display(Markdown(x))\n",
    "def dh(x): display(HTML(layout_style() + x))\n",
    "\n",
    "    \n",
    "def display_single_context(context, start_index, end_index):\n",
    "    \n",
    "    before_answer = context[:start_index]\n",
    "    answer = context[start_index:end_index]\n",
    "    after_answer = context[end_index:]\n",
    "    \n",
    "    content = before_answer + \"<span class='answer'>\" + answer + \"</span>\" + after_answer\n",
    "    \n",
    "    return dh(\"\"\"<div class=\"single_answer\">{}</div>\"\"\".format(content))\n",
    "\n",
    "def display_question_title(question):\n",
    "    return dh(\"<h2 class='question_title'>{}</h2>\".format(question)) #.capitalize()\n",
    "\n",
    "def display_all_contexts(question):\n",
    "    \n",
    "    def answer_not_found(context, start_index, end_index):\n",
    "        return (start_index == 0 and len(context) == end_index) or (start_index == 0 and end_index == 0)\n",
    "\n",
    "    display_question_title(str(1) + \". \" + question['question'].capitalize())\n",
    "    \n",
    "    # display context\n",
    "    for i in question['results']:\n",
    "        for context, ref in zip(results_answers.context, results_answers.reference):\n",
    "            if i['context'] in context:\n",
    "                \n",
    "                if answer_not_found(i['context'], i['start_index'], i['end_index']):\n",
    "                    continue # skip not found questions\n",
    "                display_single_context(i['context']+'<br>'+'<strong>'+'<font color=black>'+ref,\n",
    "                                       i['start_index'], i['end_index'])\n",
    "\n",
    "def display_task_title(task):\n",
    "    task_title = \"Task \" + str('index') + \": \" + task\n",
    "    return dh(\"<h1 class='task_title'>{}</h1>\".format(task_title))\n",
    "\n",
    "def display_single_task(task):\n",
    "    \n",
    "    display_task_title(task['task'])\n",
    "    \n",
    "    for i, question in enumerate(task['questions']):\n",
    "        display_all_contexts(question)\n",
    "        \n",
    "task = 8\n",
    "\n",
    "display_single_task(cord19_answers_json['data'][task-1])\n",
    "# =========================================================================================================\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Can 2019-nCoV infect patients a second time?',\n",
       " 'Can personal protective equipment prevent the transmission of 2019-nCoV?',\n",
       " 'What is the immune system response to 2019-nCoV?'}"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = {'data':{q['question'] for q in tq['questions']} for tq in cord19_answers_json['data']}\n",
    "q['data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>task</th>\n",
       "      <th>question</th>\n",
       "      <th>context</th>\n",
       "      <th>answer</th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>What is known about transmission, incubation, ...</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets,...</td>\n",
       "      <td>Viral infections can be transmitted by various...</td>\n",
       "      <td>transmission is via aerosols , droplets , or f...</td>\n",
       "      <td>Limeres Posse, Jacobo et al. (2017)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>What is known about transmission, incubation, ...</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets,...</td>\n",
       "      <td>Bacterial infections have a large impact on pu...</td>\n",
       "      <td>air , water , food , or living vectors</td>\n",
       "      <td>Doron, S.; Gorbach, S.L. (2008)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>What is known about transmission, incubation, ...</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets,...</td>\n",
       "      <td>BACKGROUND: Respiratory infections are the lea...</td>\n",
       "      <td>close contact with a sick person ' s breath , ...</td>\n",
       "      <td>Nizame, Fosiul A et al. (2011)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>What is known about transmission, incubation, ...</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets,...</td>\n",
       "      <td>Emerging infectious diseases (EID) and reemerg...</td>\n",
       "      <td>direct contact or through food , water , and t...</td>\n",
       "      <td>McArthur, Donna Behler (2019)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>What is known about transmission, incubation, ...</td>\n",
       "      <td>Is the virus transmitted by aerisol, droplets,...</td>\n",
       "      <td>Viruses are often transmitted via food and the...</td>\n",
       "      <td>viruses are often transmitted via food and the...</td>\n",
       "      <td>Cliver, Dean O. (2008)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>Other interesting questions</td>\n",
       "      <td>Can 2019-nCoV infect patients a second time?</td>\n",
       "      <td>AbstractThere is concern about a new coronavir...</td>\n",
       "      <td>can 2019 - ncov infect patients a second time...</td>\n",
       "      <td>Domenico Benvenuto et al. (2020)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>Other interesting questions</td>\n",
       "      <td>Can 2019-nCoV infect patients a second time?</td>\n",
       "      <td>BACKGROUND The outbreak of a novel coronavirus...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bo Diao et al. (2020)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>Other interesting questions</td>\n",
       "      <td>Can 2019-nCoV infect patients a second time?</td>\n",
       "      <td>AbstractOver the past 20 years, several corona...</td>\n",
       "      <td>human ace2 is the receptor</td>\n",
       "      <td>Michael Letko; Vincent Munster (2020)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>Other interesting questions</td>\n",
       "      <td>Can 2019-nCoV infect patients a second time?</td>\n",
       "      <td>INTRODUCTION: The outbreak of the new Coronavi...</td>\n",
       "      <td>march 18 , 2020</td>\n",
       "      <td>Panahi, Latif et al. (2020)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>Other interesting questions</td>\n",
       "      <td>Can 2019-nCoV infect patients a second time?</td>\n",
       "      <td>In December 2019, a viral pneumonia outbreak c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kim, Jin Yong et al. (2020)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  task  \\\n",
       "0    What is known about transmission, incubation, ...   \n",
       "1    What is known about transmission, incubation, ...   \n",
       "2    What is known about transmission, incubation, ...   \n",
       "3    What is known about transmission, incubation, ...   \n",
       "4    What is known about transmission, incubation, ...   \n",
       "..                                                 ...   \n",
       "295                        Other interesting questions   \n",
       "296                        Other interesting questions   \n",
       "297                        Other interesting questions   \n",
       "298                        Other interesting questions   \n",
       "299                        Other interesting questions   \n",
       "\n",
       "                                              question  \\\n",
       "0    Is the virus transmitted by aerisol, droplets,...   \n",
       "1    Is the virus transmitted by aerisol, droplets,...   \n",
       "2    Is the virus transmitted by aerisol, droplets,...   \n",
       "3    Is the virus transmitted by aerisol, droplets,...   \n",
       "4    Is the virus transmitted by aerisol, droplets,...   \n",
       "..                                                 ...   \n",
       "295       Can 2019-nCoV infect patients a second time?   \n",
       "296       Can 2019-nCoV infect patients a second time?   \n",
       "297       Can 2019-nCoV infect patients a second time?   \n",
       "298       Can 2019-nCoV infect patients a second time?   \n",
       "299       Can 2019-nCoV infect patients a second time?   \n",
       "\n",
       "                                               context  \\\n",
       "0    Viral infections can be transmitted by various...   \n",
       "1    Bacterial infections have a large impact on pu...   \n",
       "2    BACKGROUND: Respiratory infections are the lea...   \n",
       "3    Emerging infectious diseases (EID) and reemerg...   \n",
       "4    Viruses are often transmitted via food and the...   \n",
       "..                                                 ...   \n",
       "295  AbstractThere is concern about a new coronavir...   \n",
       "296  BACKGROUND The outbreak of a novel coronavirus...   \n",
       "297  AbstractOver the past 20 years, several corona...   \n",
       "298  INTRODUCTION: The outbreak of the new Coronavi...   \n",
       "299  In December 2019, a viral pneumonia outbreak c...   \n",
       "\n",
       "                                                answer  \\\n",
       "0    transmission is via aerosols , droplets , or f...   \n",
       "1               air , water , food , or living vectors   \n",
       "2    close contact with a sick person ' s breath , ...   \n",
       "3    direct contact or through food , water , and t...   \n",
       "4    viruses are often transmitted via food and the...   \n",
       "..                                                 ...   \n",
       "295   can 2019 - ncov infect patients a second time...   \n",
       "296                                                NaN   \n",
       "297                         human ace2 is the receptor   \n",
       "298                                    march 18 , 2020   \n",
       "299                                                NaN   \n",
       "\n",
       "                                 reference  \n",
       "0      Limeres Posse, Jacobo et al. (2017)  \n",
       "1          Doron, S.; Gorbach, S.L. (2008)  \n",
       "2           Nizame, Fosiul A et al. (2011)  \n",
       "3            McArthur, Donna Behler (2019)  \n",
       "4                   Cliver, Dean O. (2008)  \n",
       "..                                     ...  \n",
       "295       Domenico Benvenuto et al. (2020)  \n",
       "296                  Bo Diao et al. (2020)  \n",
       "297  Michael Letko; Vincent Munster (2020)  \n",
       "298            Panahi, Latif et al. (2020)  \n",
       "299            Kim, Jin Yong et al. (2020)  \n",
       "\n",
       "[300 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "results_answers = pd.read_csv('../data_processed/cord19_answers.csv')\n",
    "\n",
    "results_answers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_display_children(selected_task, selected_question, context):\n",
    "    ra = results_answers[(results_answers.task== selected_task) &\n",
    "                         (results_answers.question==selected_question)]['context']\n",
    "    output = {'context'+str(j):[] for j, t in enumerate(ra)}\n",
    "    return output[context]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(P(None),)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import dash_html_components as html\n",
    "\n",
    "html.P(),\n",
    "html.Div()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "270    Background China is running a national level a...\n",
       "271    The recent outbreak of SARS-CoV-2 (2019-nCoV) ...\n",
       "272    The 2019-nCoV is reported to share the same en...\n",
       "273    The outcome of a viral infection within the ne...\n",
       "274    AbstractA newly identified coronavirus, 2019-n...\n",
       "275    Timely detection of novel coronavirus (2019-nC...\n",
       "276    This paper focuses on the formulation of a det...\n",
       "277    The 2019-nCoV infection that is caused by a no...\n",
       "278    AbstractSince December 2019, a novel coronavir...\n",
       "279    ABSTRACTThe respiratory syndrome caused by a n...\n",
       "280    In late December 2019 a previous unidentified ...\n",
       "281    We evaluated a personal protective equipment r...\n",
       "282    ABSTRACT THE CENTERS FOR DISEASE CONTROL and P...\n",
       "283    The severe acute respiratory syndrome (SARS) o...\n",
       "284    Over the past several decades, we have witness...\n",
       "285    A possible threat in the ophthalmology clinic ...\n",
       "286    In late December 2019, a previous unidentified...\n",
       "287    Healthcare settings can amplify transmission o...\n",
       "288    This document outlines recommendations for phy...\n",
       "289    Summary The epidemiology and virology of sever...\n",
       "290    Radiologic characteristics of 2019 novel coron...\n",
       "291    AbstractThe outbreak of 2019-nCoV pneumonia (C...\n",
       "292    The present paper describes the mathematical m...\n",
       "293    Radiologic characteristics of 2019 novel coron...\n",
       "294    Translation of a genetic codon without a cogna...\n",
       "295    AbstractThere is concern about a new coronavir...\n",
       "296    BACKGROUND The outbreak of a novel coronavirus...\n",
       "297    AbstractOver the past 20 years, several corona...\n",
       "298    INTRODUCTION: The outbreak of the new Coronavi...\n",
       "299    In December 2019, a viral pneumonia outbreak c...\n",
       "Name: context, dtype: object"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = []\n",
    "d = results_answers[(results_answers.task.str.contains('Other interesting questions')==True) | \n",
    "                                        (results_answers.question.str.contains('Can 2019-nCoV infect patients a second time?')==True)]['context']\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'context0'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-9750c77f5439>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'context'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0md\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0moutput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'context'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'context0'"
     ]
    }
   ],
   "source": [
    "output = {'context'+str(j):[]}\n",
    "for j, out in enumerate(d):\n",
    "    output['context'+str(j)].append(out)\n",
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'todict'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-49-78f3e1c6012d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Context'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m': { '\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m' }'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtodict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'todict'"
     ]
    }
   ],
   "source": [
    "output = []\n",
    "for j, out in enumerate(d):\n",
    "    output.append('Context'+str(j)+': { '+out+' }')\n",
    "\n",
    "output.todict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Background China is running a national level antivirus campaign against the novel coronavirus (2019-nCoV). Strict control measures are being enforced in either the populated areas and remote regions. While the virus is closed to be under control, tremendous economic loss has been caused. Methods and findings We assessed the pandemic risk of 2019-nCoV for all cities/regions in China using the random forest algorithm, taking into account the effect of five factors: the accumulative and increased numbers of confirmed cases, total population, population density, and GDP. We defined four levels of the risk, corresponding to the four response levels to public health emergencies in China. The classification system has good consistency among cities in China, as the error rate of the confusion matrix is 1.58%. Conclusions The pandemic risk of 2019-nCoV is dramatically different among the 442 cities/regions. We recommend to adopt proportionate control policy according to the risk level to reduce unnecessary economic loss.'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output = {'Context'+str(j): {out} for j, out in enumerate(d)}\n",
    "output['Context0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ajjoute un element vide dans le dict pour pouvoir metre la valeur par default None ou ''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Projet Corona (Python)",
   "language": "python",
   "name": "corona"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
